src/app/InteractionModelEngine.cpp

```diff
@@ -485,6 +540,223 @@ void InteractionModelEngine::AddReadClient(ReadClient * apReadClient)
+
+#if CONFIG_IM_BUILD_FOR_UNIT_TEST
+    const int32_t pathPoolCapacity =
+        (mPathPoolCapacityOverride == -1) ? CHIP_IM_SERVER_MAX_NUM_PATH_GROUPS : static_cast<int32_t>(mPathPoolCapacityOverride);
+    const int32_t readHandlerPoolCapacity =
+        (mReadHandlerCapacityOverride == -1) ? CHIP_IM_MAX_NUM_READ_HANDLER : static_cast<int32_t>(mReadHandlerCapacityOverride);
+#else
+    const int32_t pathPoolCapacity        = CHIP_IM_SERVER_MAX_NUM_PATH_GROUPS;
+    const int32_t readHandlerPoolCapacity = CHIP_IM_MAX_NUM_READ_HANDLER;
+#endif
+
+    // If we return early here, the compiler will complain about the unreachable code, so we add a always-true check.
+    const int32_t attributePathCap = overrideResult ? INT32_MAX : (pathPoolCapacity - static_cast<int32_t>(kReservedPathsForReads));
+    const int32_t eventPathCap     = overrideResult ? INT32_MAX : (pathPoolCapacity - static_cast<int32_t>(kReservedPathsForReads));
+    const int32_t readHandlerCap =
+        overrideResult ? INT32_MAX : (readHandlerPoolCapacity - static_cast<int32_t>(kReservedHandlerForReads));
+
+    int32_t usedAttributePaths = static_cast<int32_t>(mAttributePathPool.Allocated());
+    int32_t usedEventPaths     = static_cast<int32_t>(mEventPathPool.Allocated());
+    int32_t usedReadHandlers   = static_cast<int32_t>(mReadHandlers.Allocated());
+
+    if (usedAttributePaths + aRequestedAttributePathCount <= attributePathCap &&
+        usedEventPaths + aRequestedEventPathCount <= eventPathCap && usedReadHandlers < readHandlerCap)
+    {
+        // We have enough resources, then we serve the requests in a best-effort manner.
+        return true;
+    }
+
+    if ((aRequestedAttributePathCount > static_cast<int32_t>(kMinSupportedPathPerSubscription) &&
+         usedAttributePaths + aRequestedAttributePathCount > attributePathCap) ||
+        (aRequestedEventPathCount > static_cast<int32_t>(kMinSupportedPathPerSubscription) &&
+         usedEventPaths + aRequestedEventPathCount > eventPathCap))
+    {
+        // We cannot offer enough resources, and the subscription is requesting more than the spec limit.
+        return false;
+    }
+
+    bool hasEvictedHandlers = true;
+
+    const auto evictAndUpdateResourceUsage = [&](FabricIndex fabricIndex, bool forceEvict) {
+        bool ret           = CheckResourceAndEvictExceededSubscription(fabricIndex, forceEvict);
+        usedAttributePaths = static_cast<int32_t>(mAttributePathPool.Allocated());
+        usedEventPaths     = static_cast<int32_t>(mEventPathPool.Allocated());
+        usedReadHandlers   = static_cast<int32_t>(mReadHandlers.Allocated());
+        return ret;
+    };
+
+    while ((usedAttributePaths + aRequestedAttributePathCount > attributePathCap ||
+            usedEventPaths + aRequestedEventPathCount > eventPathCap || usedReadHandlers >= readHandlerCap) &&
+           hasEvictedHandlers)
+    {
+        hasEvictedHandlers = false;
+        for (const auto & fabric : *mpFabricTable)
```

Comment:
We shouldn't need the outer while loop here right?

We have a compliant request that needs space and we need to kick out a non-compliant handler. Which means that a single iteration through all fabrics and progressively 'trimming' them down to be compliant should be sufficient. And if it isn't, it means we need to eject a subscription from the accessing fabric right?

Reply:
What if the incoming request has 3 paths, while one fabric has lots of subscriptions with one path?

Reply:
Gotcha. So, why not have the outer while loop be `while(didEvictHandler)`? The inner loop should at least evict one handler, and if we hit the outer loop, it means that we still have overloaded fabrics that need trimming, and should have at least evicted one handler in that process.

Reply:
What if we exactly used all resources? If we did not evict any subscription but we don't have any more resoureces, that is the case for the loop below.